packages/driver-typescript/src/TypeScriptDriver.ts

```diff
@@ -41,4 +56,138 @@ export default class TypeScriptDriver extends Driver<TypeScriptConfig> {
       rimraf.sync(path.resolve(outDir));
     }
   };
+
+  /**
+   * Define references and compiler options when `--reference-workspaces` option is passed.
+   */
+  handleCreateConfig = (
+    context: DriverContext<DriverArgs & TypeScriptArgs & { referenceWorkspaces?: boolean }>,
+    configPath: string,
+    config: TypeScriptConfig,
+  ) => {
+    const { args, workspaceRoot } = context;
+
+    if (!args.referenceWorkspaces) {
+      return;
+    }
+
+    // Extract compiler options to a new config file
+    const optionsPath = path.join(path.dirname(configPath), 'tsconfig.options.json');
+
+    fs.writeFileSync(
+      optionsPath,
+      this.formatConfig({
+        compilerOptions: {
+          ...config.compilerOptions,
+          // Required for project references
+          composite: true,
+          declaration: true,
+          declarationMap: true,
+          // Remove by marking as undefined
+          outDir: undefined,
+          outFile: undefined,
+        },
+      }),
+    );
+
+    delete config.compilerOptions;
+
+    // Generate references and update paths
+    config.extends = './tsconfig.options.json';
+    config.files = [];
+    config.references = this.tool.getWorkspacePackages({ root: workspaceRoot }).map(wsPkg => ({
+      path: path.relative(workspaceRoot, wsPkg.workspace.packagePath),
+    }));
```

Comment:
```ts
    config.references = []
        
    const { testsFolder } = this.options;
    this.tool.getWorkspacePackages({ root: workspaceRoot }).forEach(wsPkg => {
      config.references.push({ path: path.relative(workspaceRoot, wsPkg.workspace.packagePath) })
      
      if (testsFolder) {
         const testsPath = path.join(wsPkg.workspace.packagePath, testsFolder);
         
         if (fs.existsSync(testsPath) {
           config.references.push({ path: path.relative(workspaceRoot, testsPath) })
         }
    })
```

Reply:
This works unless a package declares custom type declaration files, which are then outside of the tests "root". >_<

Like here: https://github.com/beemojs/beemo/tree/master/packages/core/types/hygen

Trying to figure this out now.

Reply:
I also seem to hit this a lot... Which is unfortunate.

```
/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25865
                ts.Debug.assert(source !== target);
                         ^
Error: Debug Failure. False expression.
    at mergeSymbol (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25865:26)
    at /Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25964:47
    at Map.forEach (<anonymous>)
    at mergeSymbolTable (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25962:20)
    at mergeSymbol (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25887:21)
    at /Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25964:47
    at Map.forEach (<anonymous>)
    at mergeSymbolTable (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25962:20)
    at mergeSymbol (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25892:21)
    at /Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25964:47
```

Reply:
Actually, it was the config upstream in `@milesj` causing it. However, builds are much slower including tests also.

Reply:
thats lame...